vendor/bat-native-ledger/src/bat/ledger/internal/contribution/contribution_unblinded.cc

```diff
@@ -471,18 +596,77 @@ void Unblinded::Retry(
     return;
   }
 
-  if (contribution->step == ledger::ContributionStep::STEP_START) {
-    Start(types, contribution->contribution_id, callback);
-    return;
+  switch (contribution->step) {
+    case ledger::ContributionStep::STEP_START: {
+      Start(types, contribution->contribution_id, callback);
+      return;
+    }
+    case ledger::ContributionStep::STEP_PREPARE: {
+      ProcessTokens(types, contribution->contribution_id, callback);
+      return;
+    }
+    case ledger::ContributionStep::STEP_RESERVE: {
+      // Retrieve reserved unblinded tokens from cache for retry
+      // attempt
+      std::vector<ledger::UnblindedToken> list;
+      if (!GetReservedUnblindedTokensFromCache(
+            contribution->contribution_id,
+            &list)) {
+        BLOG(0, "No reserved unblinded tokens saved for retry attempt");
+        callback(ledger::Result::LEDGER_ERROR);
+        return;
+      }
+
+      // Discard reserved unblinded tokens from cache
+      DiscardReservedUnblindedTokensFromCache(contribution->contribution_id);
+
+      PreparePublishers(list, std::move(contribution), types, callback);
+      return;
+    }
+    case ledger::ContributionStep::STEP_AC_TABLE_EMPTY:
+    case ledger::ContributionStep::STEP_CREDS:
+    case ledger::ContributionStep::STEP_EXTERNAL_TRANSACTION:
+    case ledger::ContributionStep::STEP_NOT_ENOUGH_FUNDS:
+    case ledger::ContributionStep::STEP_FAILED:
+    case ledger::ContributionStep::STEP_COMPLETED:
+    case ledger::ContributionStep::STEP_NO: {
+      BLOG(0, "Step not correct " << contribution->step);
+      NOTREACHED();
+      return;
+    }
   }
+}
 
-  if (contribution->step == ledger::ContributionStep::STEP_PREPARE) {
-    ProcessTokens(types, contribution->contribution_id, callback);
-    return;
+void Unblinded::AddReservedUnblindedTokensToCache(
+    const std::string& contribution_id,
+    const std::vector<ledger::UnblindedToken>& list) {
+  DCHECK(!contribution_id.empty());
+  base::AutoLock lock(reserved_unblinded_tokens_lock_);
```

Comment:
I don't think I've seen other attempts to make member variable access thread-safe in `bat-native-ledger`. Do we use multiple threads/chromium sequences/etc. in `bat-native-ledger`? Or are we proactively using locks here?

Reply:
It was out of an abundance of caution, but I can remove it if it's not really needed.

Reply:
Gotcha. Up to you - it would be nice to have some design docs explaining how threading is used (or not) within shared lib to guide choices like this. (It seems that with how the code is currently structured we assume no multi-threading.)